src/app/components/IdentityProviderWizard/Wizards/Providers/Cloudflare/SAML/index.tsx

```diff
@@ -0,0 +1,254 @@
+  const [error, setError] = useState<null | boolean>(null);
+  const [isFormValid, setIsFormValid] = useState(false);
+  const [isValidating, setIsValidating] = useState(false);
+  const [disableButton, setDisableButton] = useState(false);
+  const { getRealm } = useKeycloakAdminApi();
+  const {
+    alias,
+    setAlias,
+    adminLinkSaml: adminLink,
+    loginRedirectURL: acsUrl,
+    entityId,
+    identifierURL,
+    createIdPUrl,
+    baseServerRealmsUrl,
+  } = useApi();
+
+  const [metadata, setMetadata] = useState<METADATA_CONFIG>();
+  const [metadataUrl, setMetadataUrl] = useState("");
+
+  useEffect(() => {
+    const genAlias = getAlias({
+      provider: Providers.CLOUDFLARE,
+      protocol: Protocols.SAML,
+      preface: "cloudflare-saml",
+    });
+    setAlias(genAlias);
+  }, []);
+
+  const finishStep = 6;
+
+  usePrompt(
+    "The wizard is incomplete. Leaving will lose any saved progress. Are you sure?",
+    stepIdReached < finishStep
+  );
+
+  const onNext = (newStep) => {
+    if (stepIdReached === finishStep) {
+      clearAlias({
+        provider: Providers.CLOUDFLARE,
+        protocol: Protocols.SAML,
+      });
+      navigateToBasePath();
+    }
+    setStepIdReached(stepIdReached < newStep.id ? newStep.id : stepIdReached);
+  };
+
+  const closeWizard = () => {
+    navigateToBasePath();
+  };
+
+  const handleFormSubmit = async ({
+    url,
+  }: {
+    url: string;
+  }): Promise<API_RETURN> => {
+    // make call to submit the URL and verify it
+    setMetadataUrl(metadataUrl);
+
+    try {
+      const payload = {
+        fromUrl: url,
+        providerId: "saml",
+        realm: getRealm(),
+      };
+      const resp = await Axios.post(identifierURL, payload);
+
+      if (resp.status === 200) {
+        setMetadata({
+          ...SamlIDPDefaults,
+          ...resp.data,
+        });
+        setIsFormValid(true);
+        return {
+          status: API_STATUS.SUCCESS,
+          message: `Configuration successfully validated with ${idpCommonName}. Continue to next step.`,
+        };
+      }
+    } catch (err) {
+      console.log(err);
+    }
+
+    return {
+      status: API_STATUS.ERROR,
+      message: `Configuration validation failed with ${idpCommonName}. Check file and try again.`,
+    };
+  };
+
+  const createCloudflareSamlIdP = async () => {
+    // On final validation set stepIdReached to steps.length+1
+    setIsValidating(true);
+    setDisableButton(false);
+    setResults(`Creating ${idpCommonName}...`);
+
+    const payload: IdentityProviderRepresentation = {
+      alias,
+      displayName: `Cloudflare SAML Single Sign-on`,
+      providerId: "saml",
+      config: metadata!,
+    };
+
+    try {
+      await CreateIdp({ createIdPUrl, payload, featureFlags });
+
+      await SamlAttributeMapper({
+        alias,
+        createIdPUrl,
+        usernameAttribute: { attributeName: "email", friendlyName: "" }, //using email for username
```

Comment:
@cdgco i assume this is due to how Cloudflare doesn't have a "username" value?

Reply:
Cloudflare is a bit weird as an IdP as it's really a passthrough to whatever IdPs are configured for the account, and the attributes can vary user to user as you can have users from different source IdPs in the same Cloudflare account. Depending on the IdP, it might not provide a username, but it should always provide an email.

Reply:
You can check the attributes under Users -> [User] -> User Registry identity -> Get-identity endpoint. `Email` is a system attribute while things like username would fall under the custom `oidc_fields` for an OAuth source provider.